internal/cluster/connection.go

```diff
@@ -0,0 +1,256 @@
+)
+
+const (
+	bufferSize      = 8 * 1024
+	protocolStarter = "CP2"
+)
+
+const (
+	open int32 = iota
+	closed
+)
+
+type ResponseHandler func(msg *proto.ClientMessage)
+
+type Connection struct {
+	responseCh                chan<- *proto.ClientMessage
+	pending                   chan *proto.ClientMessage
+	received                  chan *proto.ClientMessage
+	socket                    net.Conn
+	doneCh                    chan struct{}
+	endpoint                  atomic.Value
+	status                    int32
+	lastRead                  atomic.Value
+	lastWrite                 atomic.Value
+	closedTime                atomic.Value
+	writeBuffer               []byte
+	connectionID              int64
+	eventDispatcher           *event.DispatchService
+	connectedServerVersion    int32
+	connectedServerVersionStr string
+	startTime                 int64
+	logger                    ilogger.Logger
+}
+
+func (c *Connection) ConnectionID() int64 {
+	return c.connectionID
+}
+
+func (c *Connection) start(clusterCfg *pubcluster.Config, addr pubcluster.Address) error {
+	if socket, err := c.createSocket(clusterCfg, addr); err != nil {
+		return err
+	} else {
+		c.socket = socket
+		c.lastWrite.Store(time.Time{})
+		c.closedTime.Store(time.Time{})
+		c.startTime = timeutil.GetCurrentTimeInMilliSeconds()
+		c.lastRead.Store(time.Now())
+		if err := c.sendProtocolStarter(); err != nil {
+			c.socket.Close()
+			c.socket = nil
+			return err
+		}
+		go c.socketReadLoop()
+		go c.socketWriteLoop()
+		return nil
+	}
+}
+
+func (c *Connection) sendProtocolStarter() error {
+	_, err := c.socket.Write([]byte(protocolStarter))
+	return err
+}
+
+func (c *Connection) createSocket(networkCfg *pubcluster.Config, address pubcluster.Address) (net.Conn, error) {
+	conTimeout := timeutil.GetPositiveDurationOrMax(networkCfg.ConnectionTimeout)
+	socket, err := c.dialToAddressWithTimeout(address, conTimeout)
+	if err != nil {
+		return nil, err
+	}
+	//if networkCfg.SSLConfig().Enabled() {
+	//	socket, err = c.openTLSConnection(networkCfg.SSLConfig(), socket)
+	//}
+	return socket, err
+}
+
+func (c *Connection) dialToAddressWithTimeout(addr pubcluster.Address, conTimeout time.Duration) (*net.TCPConn, error) {
+	if conn, err := net.DialTimeout("tcp", addr.String(), conTimeout); err != nil {
+		return nil, err
+	} else {
+		tcpConn := conn.(*net.TCPConn)
+		if err = tcpConn.SetNoDelay(true); err != nil {
+			panic(fmt.Errorf("error setting tcp no delay: %w", err))
+		}
+		if err = tcpConn.SetReadBuffer(bufferSize); err != nil {
+			panic(fmt.Errorf("error setting read buffer: %w", err))
+		}
+		if err = tcpConn.SetWriteBuffer(bufferSize); err != nil {
+			panic(fmt.Errorf("error setting write buffer: %w", err))
+		}
+		return tcpConn, nil
+	}
+}
+
+func (c *Connection) isAlive() bool {
+	return atomic.LoadInt32(&c.status) == open
+}
+
+func (c *Connection) socketWriteLoop() {
+	for {
+		select {
+		case request, ok := <-c.pending:
```

Comment:
nit: it should be more or less trivial to implement autopipelining (Node.js client already [has it](https://github.com/hazelcast/hazelcast-nodejs-client/pull/481)) with current design, so I'm leaving some hints here.

`pending` channel's size should be increase from the current size of 1 message to something more significant (say, 64). Then when the `socketWriteLoop()` loop reads a message from the channel, it should append it to an intermediate array and then schedule a short-living timer (say, 1ms), if it's not yet scheduled. Once the timer is triggered, all accumulated messages should be written into an intermediate, reusable buffer and only then written into the socket.

Autopipelining should improve client's throughput (in Node.js client it was something 20-30% improvement in a number of benchmarks) at the cost of slightly increased latency of individual messages. For high-priority messages, like heartbeats, we could introduce a field on `ClientMessage` struct that would instruct the `socketWriteLoop()` loop to immediately write any accumulated messages.